lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -1254,67 +1199,77 @@ def determine_group_level_monitored_resource_and_labels(tag,
       # "container.<container_id>" // Docker container.
       # "k8s_pod.<namespace_name>.<pod_name>" // GKE pod.
       if local_resource_id
-        converted_resource = monitored_resource_from_local_resource_id(
-          local_resource_id)
-        resource = converted_resource if converted_resource
+        resource = monitored_resource_from_local_resource_id(local_resource_id)
+        if resource
+          common_labels = get_common_labels_for_resource_type(resource.type)
+
+          resource.freeze
+          resource.labels.freeze
+          common_labels.freeze
+          return [resource, common_labels]
+        end
       end
 
-      # Once the resource type is settled down, determine the labels.
-      case resource.type
-      # Cloud Functions.
-      when CLOUDFUNCTIONS_CONSTANTS[:resource_type]
-        resource.labels.merge!(
-          'region' => @gcf_region,
-          'function_name' => decode_cloudfunctions_function_name(
-            matched_regexp_group['encoded_function_name'])
-        )
-        instance_id = resource.labels.delete('instance_id')
-        common_labels.merge!(
-          "#{GKE_CONSTANTS[:service]}/instance_id" => instance_id,
-          "#{COMPUTE_CONSTANTS[:service]}/resource_id" => instance_id,
-          "#{GKE_CONSTANTS[:service]}/cluster_name" =>
-            resource.labels.delete('cluster_name'),
-          "#{COMPUTE_CONSTANTS[:service]}/zone" =>
-            resource.labels.delete('zone')
-        )
+      resource_type = determine_agent_level_monitored_resource_type
+      resource = Google::Apis::LoggingV2::MonitoredResource.new(
+        type: resource_type,
+        labels: determine_agent_level_monitored_resource_labels(
```

Comment:
Could we not create a `discover_monitored_resource` function that actually constructs the resource (i.e., resurrect `determine_agent_level_monitored_resource_via_legacy` under a new name)? We can then inline the functions for the type and the labels.